File: MiniDOMOffice/Office/Uscite/LuoghiVisitati/LuoghiVisitatiPerUscitaCollection.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Office;


namespace minidom
{
    public partial class Office
    {

        /// <summary>
        /// <see cref="Luogo"/> visitato durante una <see cref="Uscita"/>
        /// </summary>
        [Serializable]
        public class LuoghiVisitatiPerUscitaCollection
            : CCollection<LuogoVisitato>
        {
            [NonSerialized] private Uscita m_Uscita;

            /// <summary>
            /// Costruttore
            /// </summary>
            public LuoghiVisitatiPerUscitaCollection()
            {
                m_Uscita = null;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="uscita"></param>
            public LuoghiVisitatiPerUscitaCollection(Uscita uscita)
            {
                if (uscita is null)
                    throw new ArgumentNullException("uscita");
                Load(uscita);
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, LuogoVisitato value)
            {
                if (m_Uscita is object)
                    value.SetUscita(m_Uscita);
                base.OnInsert(index, value);
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, LuogoVisitato oldValue, LuogoVisitato newValue)
            {
                if (m_Uscita is object)
                    newValue.SetUscita(m_Uscita);
                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// Carica
            /// </summary>
            /// <param name="uscita"></param>
            protected internal void Load(Uscita uscita)
            {
                if (uscita is null)
                    throw new ArgumentNullException("uscita");
                
                this.Clear();
                this.m_Uscita = uscita;
                
                if (DBUtils.GetID(uscita, 0) == 0)
                    return;

                using (var cursor = new LuoghiVisitatiCursor())
                {
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    cursor.OraArrivo.SortOrder = SortEnum.SORT_ASC;
                    cursor.IDUscita.Value = DBUtils.GetID(uscita, 0);
                    cursor.Progressivo.SortOrder = SortEnum.SORT_ASC;
                    cursor.IgnoreRights = true;
                    while (cursor.Read())
                    {
                        Add(cursor.Item);
                    }
                }
                     
            }
        }
    }
}

File: MiniDOMOffice/Office/LuoghiVisitati/LuoghiVisitatiPerUscitaCollection.vb
Imports minidom.Databases
Imports minidom.Sistema


Partial Class Office

    <Serializable> _
    Public Class LuoghiVisitatiPerUscitaCollection
        Inherits CCollection(Of LuogoVisitato)

        Private m_Uscita As Uscita

        Public Sub New()
            Me.m_Uscita = Nothing
        End Sub

        Public Sub New(ByVal uscita As Uscita)
            If (uscita Is Nothing) Then Throw New ArgumentNullException("uscita")
            Me.Load(uscita)
        End Sub

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Uscita IsNot Nothing) Then DirectCast(value, LuogoVisitato).SetUscita(Me.m_Uscita)
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Uscita IsNot Nothing) Then DirectCast(newValue, LuogoVisitato).SetUscita(Me.m_Uscita)
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Protected Friend Sub Load(ByVal uscita As Uscita)
            If (uscita Is Nothing) Then Throw New ArgumentNullException("uscita")
            Me.Clear()
            Me.m_Uscita = uscita
            If (GetID(uscita) = 0) Then Exit Sub
            Dim cursor As New LuoghiVisitatiCursor
            Try
                cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                cursor.OraArrivo.SortOrder = SortEnum.SORT_ASC
                cursor.IDUscita.Value = GetID(uscita)
                cursor.Progressivo.SortOrder = SortEnum.SORT_ASC
                cursor.IgnoreRights = True
                While Not cursor.EOF
                    Me.Add(cursor.Item)
                    cursor.MoveNext()
                End While
            Catch ex As Exception
                Sistema.Events.NotifyUnhandledException(ex)
                Throw
            Finally
                If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing
            End Try
        End Sub

    End Class



End Class